### ISSUE TITLE
MQTT AD missing call to create sensors

### ISSUE THREAD

**Issue by redswan01** [domoticz/domoticz#5280]
Hi,
I have been trying to debug why my Sonoff SNZB-02 were not being created despite being discovered. I have come to the conclusion that there's a missing call in the function "on_auto_discovery_message" in MqTTAutodiscover.cpp at line 1046/7, 
`		if (pSensor->component_type == "sensor")
		{
			if (pSensor->state_topic.empty())
			{
				Log(LOG_ERROR, "A sensor should have a state_topic!");
				return;
			}
			handle_auto_discovery_sensor(pSensor, message);  // This line is missing
		}
`
After adding this line then devices are created, though in my case I get individual devices for for temperature, humidity and battery as opposed to one combined temp/humidity device.

Maybe someone with more knowledge of the intricacies could confirm this is the correct place to do this, and also make a pr for me as I'm not even a novice where github is concerned.

I first came across this problem with 2022.1  build 14352 on a Rpi 3b with Bullseye, to debug I cloned 2022.1 build 14394 on windows and built/debugged with VS2022.

Thanks.

Chris

**Comment by gizmocuz** [comment]
Sensors are 'detected' and 'created' when actually received
Only setpoints/lights/switches are created when detected

**Comment by redswan01** [comment]
OK, thanks for the explanation, that explains why I thought it was missing.
However, in may case, this does not happen, neither for the Sonoff SNZB-02 nor for temp/humidity sensor I try to create following the example in the HomeAssitant doc, I've added a unique_ID so that Domoticz doesn't create a new one at each discovery.
Both these sensors work correctly in HA, in fact for the moment I use HA to forward the data to Domoticz!!
Apparently Domoticz discovers these sensors, but sending data doesn't create any devices!

2022-07-21 11:23:47.762  Status: MQTT broker: discovered: Temp_Sensor_1/Temp_Sensor_1_temperature (unique_id: 0x00124b00251ce5b2_temperature_homeassistant)
2022-07-21 11:23:47.870  Status: MQTT broker: discovered: Temp_Sensor_1/Temp_Sensor_1_humidity (unique_id: 0x00124b00251ce5b2_humidity_homeassistant)
2022-07-21 11:23:47.984  Status: MQTT broker: discovered: Temp_Sensor_1/Temp_Sensor_1_voltage (unique_id: 0x00124b00251ce5b2_voltage_homeassistant)
2022-07-21 11:23:48.199  Status: MQTT broker: discovered: 1234_temperature_homeassistant/Temperature (unique_id: 1234_temperature_homeassistant)
2022-07-21 11:23:48.310  Status: MQTT broker: discovered: 1234_humidity_homeassistant/Humidity (unique_id: 1234_humidity_homeassistant)

Here are the config and data topics and payloads for the above:

SNZB-02:
config topic : homeassistant/sensor/0x00124b00251ce5b2/temperature/config
payload : 
{
    "availability": [{
            "topic": "homeassistant/bridge/state",
            "value_template": "{{ value_json.state }}"
        }
    ],
    "device": {
        "identifiers": ["zigbee2mqtt_0x00124b00251ce5b2"],
        "manufacturer": "SONOFF",
        "model": "Temperature and humidity sensor (SNZB-02)",
        "name": "Temp_Sensor_1"
    },
    "device_class": "temperature",
    "enabled_by_default": true,
    "name": "Temp_Sensor_1_temperature",
    "state_class": "measurement",
    "state_topic": "homeassistant/Temp_Sensor_1",
    "unique_id": "0x00124b00251ce5b2_temperature_homeassistant",
    "unit_of_measurement": "°C",
    "value_template": "{{ value_json.temperature }}"
}

for Humidity, Battery, Voltage and Link quality the schema is the same.

data topic : homeassistant/Temp_Sensor_1

payload : {"battery":100,"humidity":54.36,"linkquality":30,"temperature":28.08,"voltage":3200}

HA example:

Config topic : homeassistant/sensor/sensorBedroomT/config
payload : {
  "device_class": "temperature",
  "name": "Temperature",
  "unique_id": "1234_temperature_homeassistant",
  "state_topic": "homeassistant/sensor/sensorBedroom/state",
  "unit_of_measurement": "°C",
  "value_template": "{{ value_json.temperature}}"
}

data topic : homeassistant/sensor/sensorBedroom/state
payload : { "temperature": 23.20, "humidity": 43.70 }

Following what happens the SNZB-02 fails at this point in the function "on_auto_discovery_message"

	if (strarray.size() == 1)
	{
		//could be offline/online
		return;
	}

as the array size == 1: wrong topic format?

the HA example fails here 

	if (action != "config")
	{
		if (action == "state")
		{
			//0/1 (or online/offline)
			return;
		}

as the action is "state"

Following what you said and looking at the code I would have expected the incoming data to treated by the function "handle_auto_discovery_sensor_message" rather than "on_auto_discovery_message".

Can you help me understand why these sensors work correctly in HA and not in Domoticz.

BTW I have posted a couple of questions in the forum abut this, see user Redswan, but never got any replies! Hence my attempts to try debugging.

**Comment by redswan01** [comment]
Maybe I've found the answer

I changed the order of the two checks so that the check for an already subscribed topic exists and now it works;))
In both case the devices are created (and update) correctly.

Does this seem correct to you?

**Comment by redswan01** [comment]
Oops, forgot to include this
Modified code:

		if (m_subscribed_topics.find(topic) != m_subscribed_topics.end())
		{
			handle_auto_discovery_sensor_message(message);
			return;
		}
		if (topic.substr(0, topic.find('/')) == m_TopicDiscoveryPrefix)
		{
			on_auto_discovery_message(message);
			return;
		}

Maybe this upsets things for setpoints/lights/switches though, I'll try to test that.

**Comment by redswan01** [comment]
I have tested a few other types of devices, switch, binary sensor, light, and they all appear to function. I don't physically have these devices, I used MQTTexplorer to simulate them. I wanted to test a set-point but haven't so far found an example.

**Comment by gizmocuz** [comment]
@redswan01, I think you could be right here. This is caused because your payload goes into the same topic as the mqtt auto discovery topic.
Normally the payload goes into a topic like zwavejs2mqtt/,,,, zigbee2mqtt/.... myhardware/.....

But it won't heard to swap the two checks above, might even be faster when discovery is done 😊

Thanks for reporting! Should be solved in beta 14415

**Comment by redswan01** [comment]
@gizmocuz, I've installed the latest docker with 14415 on my Rpi and can confirm that it now works as expected for the Sonoff sensor.
As to where the payload goes: for the Sonoff sensor maybe I've miss-understood  something in the Zigbee2MQTT config file, I didn't see anything to differentiate between the config topic and the data, for my test devices I based them on the HA documentation examples - maybe this is not the correct source of information.

**Comment by gizmocuz** [comment]
@redswan01 , Good to hear ;)

For the payload, you could do something like this

`"state_topic": "myhardware/Temp_Sensor_1"`

now you need to publish to

 **myhardware/Temp_Sensor_1**

`payload : {"battery":100,"humidity":54.36,"linkquality":30,"temperature":28.08,"voltage":3200}`

**Comment by redswan01** [comment]
@gizmocuz , OK I get that, but, for the Sonoff sensor the topics used come directly from Zigbee2MqTT, so I guess that it's at this stage they are defined. If it's necessary to modify some configuration to get the correct topics then it beats the point of having auto-discovery which, if I understand correctly, should avoid the need to. Maybe the way this is configured, I suppose that somewhere there's a definition of how to translate Zigbee to MQTT, is a point to take up with the Zigbee2MqTT developer(s). Though it's probably a mute point now, as with this patch it works.

Thanks for your time.

**Comment by gizmocuz** [comment]
the payload and status for zigbee devices is usually underneat the topic zigbee2mqtt/*
maybe I don't understand 100% what you are trying to do, but as mentioned above, usually the sensors state/payload resides in a special topic/folder (not the auto discovery folder), like zwavejs2mqtt/... or zigbee2mqtt/... or tasmota/...
I think it is also a bad design to have sensors post state/payload in the auto discovery topic/folder, it will get messy and thats not what you want 😁

**Comment by redswan01** [comment]
I agree with what you're saying about separating things.
What am I trying to do: connect a Sonoff sensor to domoticz via zigbee2mqtt and mosquitto mqtt broker.
This chain works but initially, before this mod, what domoticz sees is not what it expects because the payload topics are not correct.
So the question is who defines these topics, zigbee2mqtt?
What do I need to do to get zigbee2mqtt to deliver what domoticz expects. I naively thought that this would automatically be the case, maybe I've missed something in configuring zigbee2mqtt.

**Comment by gizmocuz** [comment]
Zigbee2mqtt, you can configure it.. .maybe you configured the auto discovery topic as well as the publish topic to be the same ?
The 'base topic' in zigbee2mqtt is 'zigbee2mqtt'

**Comment by gizmocuz** [comment]
![image](https://user-images.githubusercontent.com/2265299/180451852-7de06bf4-881a-48f6-a1d5-1f8524bc6102.png)

**Comment by redswan01** [comment]
@gizmocuz, So the solution was that simple!!! I didn't realize that setting "homeassistant: true" was setting the 'auto discovery topic' and that this was different from the 'base topic', my bad! I've now changed the base topic to zigbee2mqtt and the sensors are discovered correctly and created and the payload goes into correct category.
Sooo sorry to have taken your valuable time to get here, but I've learned a lot along the way about github, building and debugging. Maybe the mod was positive anyway, even if not really necessary.
